### ISSUE TITLE
Unable to Auto Update Agent from behind a proxy

### ISSUE THREAD

**Issue by Zaniarr** [Microsoft/vsts-agent#1240]
## Have you try trouble shooting doc?
Link to trouble shooting doc: https://www.visualstudio.com/en-us/docs/build/troubleshooting
**yes**

## Agent version and platform
**- Agent.Version | 2.119.1**

Agent is running on Windows 7

## VSTS type and version
VisualStudio.com (vsts)

## What's not working?
I have an on-premise agent connected to VSTS. However, when I try and update the agent (by clicking on "Update Agent" in VSTS agent pools settings area) it does **not** update.

From the logs below it seems like there is an authentication issue. But the agent has already successfully been registered registered with VSTS. Is it trying to authenticate again to download the new version of the agent?

- I registered the agent using a PAT and the agent is running as a service using ADD credentials.
- The agent is behind a corporate proxy, I have a .proxy file at the root of the agent. The proxy is been configured to allow.
- The .proxy file only contains the proxy address i.e http://myproxy.companyname.com:9090
- The Scope of my PAT was set to Global.
- I am able to successfully run a build and release.

###Key lines from the log
````[2017-10-11 13:51:12Z WARN VisualStudioServices] Authentication failed with status code 401.
X-TFS-ServiceError: TF400813%3A%20Resource%20not%20available%20for%20anonymous%20access.%20Client%20authentication%20required.
[2017-10-11 13:51:12Z WARN VisualStudioServices] Authentication failed with status code 401.
X-TFS-ServiceError: TF400813%3A%20Resource%20not%20available%20for%20anonymous%20access.%20Client%20authentication%20required.


[2017-10-11 12:45:47Z INFO Terminal] WRITE LINE: Connecting to the server.
[2017-10-11 12:45:47Z INFO MessageListener] Attempt to create session.
[2017-10-11 12:45:47Z INFO MessageListener] Connecting to the Agent Server...
[2017-10-11 12:45:47Z INFO VisualStudioServices] Starting operation Location.GetConnectionData
[2017-10-11 12:45:49Z WARN VisualStudioServices] Authentication failed with status code 401.
Date: Wed, 11 Oct 2017 12:45:48 GMT
P3P: CP="CAO DSP COR ADMa DEV CONo TELo CUR PSA PSD TAI IVDo OUR SAMi BUS DEM NAV STA UNI COM INT PHY ONL FIN PUR LOC CNT"
Server: Microsoft-IIS/10.0
WWW-Authenticate: Bearer authorization_uri=https://login.microsoftonline.com/f9daae82-4727-4163-93b9-2214828e3dba, Basic realm="https://tfsprodweu2.app.visualstudio.com/", TFS-Federated
X-TFS-ProcessId: 75e12aa5-f063-480f-9585-cc4044c735c6
Strict-Transport-Security: max-age=31536000; includeSubDomains
ActivityId: 905a1bb5-5142-4345-9785-65d87c166b22
X-TFS-Session: f13f2eea-40f4-42de-a7e2-5875d2a1f39d
X-VSS-E2EID: 526c925f-2e15-4c52-b850-f9c47aef4d9b
X-FRAME-OPTIONS: SAMEORIGIN
X-TFS-FedAuthRealm: https://tfsprodweu2.app.visualstudio.com/
X-TFS-FedAuthIssuer: https://dddev01.visualstudio.com/
X-VSS-AuthorizationEndpoint: https://dddev01.vssps.visualstudio.com/
X-VSS-ResourceTenant: f9daae82-4727-4163-93b9-2214828e3dba
X-TFS-SoapException: %3C%3Fxml%20version%3D%221.0%22%20encoding%3D%22utf-8%22%3F%3E%3Csoap%3AEnvelope%20xmlns%3Asoap%3D%22http%3A%2F%2Fwww.w3.org%2F2003%2F05%2Fsoap-envelope%22%3E%3Csoap%3ABody%3E%3Csoap%3AFault%3E%3Csoap%3ACode%3E%3Csoap%3AValue%3Esoap%3AReceiver%3C%2Fsoap%3AValue%3E%3Csoap%3ASubcode%3E%3Csoap%3AValue%3EUnauthorizedRequestException%3C%2Fsoap%3AValue%3E%3C%2Fsoap%3ASubcode%3E%3C%2Fsoap%3ACode%3E%3Csoap%3AReason%3E%3Csoap%3AText%20xml%3Alang%3D%22en%22%3ETF400813%3A%20Resource%20not%20available%20for%20anonymous%20access.%20Client%20authentication%20required.%3C%2Fsoap%3AText%3E%3C%2Fsoap%3AReason%3E%3C%2Fsoap%3AFault%3E%3C%2Fsoap%3ABody%3E%3C%2Fsoap%3AEnvelope%3E
X-TFS-ServiceError: TF400813%3A%20Resource%20not%20available%20for%20anonymous%20access.%20Client%20authentication%20required.
X-VSS-S2STargetService: 00000002-0000-8888-8000-000000000000/visualstudio.com
X-TFS-FedAuthRedirect: https://app.vssps.visualstudio.com/_signin?realm=dddev01.visualstudio.com&reply_to=https%3A%2F%dddev01.visualstudio.com%2F_apis%2FconnectionData%3FconnectOptions%3D1%26lastChangeId%3D-1%26lastChangeId64%3D-1&redirect=1&context=eyJodCI6MiwiaGlkIjoiYTBhMmQzZDItZTcxNS00YzMyLThmZDctYTVjZjE3MjJkYWU0IiwicXMiOnt9LCJyciI6IiIsInZoIjoiIiwiY3YiOiIiLCJjcyI6IiJ90#ctx=eyJTaWduSW5Db29raWVEb21haW5zIjpbImh0dHBzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbSIsImh0dHBzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbSJdfQ2
X-Powered-By: ASP.NET
X-Content-Type-Options: nosniff
[2017-10-11 12:45:49Z INFO RSAEncryptedFileKeyManager] Loading RSA key parameters from file C:\Users\ZAhmadi\Downloads\vsts-agent-win7-x64-2.119.1\.credentials_rsaparams
[2017-10-11 12:45:49Z INFO RSAEncryptedFileKeyManager] Loading RSA key parameters from file C:\Users\ZAhmadi\Downloads\vsts-agent-win7-x64-2.119.1\.credentials_rsaparams
[2017-10-11 12:45:50Z INFO VisualStudioServices] Finished operation Location.GetConnectionData
[2017-10-11 12:45:50Z INFO MessageListener] Session created.
[2017-10-11 12:45:50Z INFO Terminal] WRITE LINE: 2017-10-11 12:45:50Z: Listening for Jobs
[2017-10-11 13:16:02Z INFO MessageListener] No message retrieved from session '71e216c8-f6dc-4e56-9648-76d09c99ce6b' within last 30 minutes.
[2017-10-11 13:46:08Z INFO MessageListener] No message retrieved from session '71e216c8-f6dc-4e56-9648-76d09c99ce6b' within last 30 minutes.
[2017-10-11 13:51:12Z WARN VisualStudioServices] Authentication failed with status code 401.
Date: Wed, 11 Oct 2017 13:51:12 GMT
P3P: CP="CAO DSP COR ADMa DEV CONo TELo CUR PSA PSD TAI IVDo OUR SAMi BUS DEM NAV STA UNI COM INT PHY ONL FIN PUR LOC CNT"
Server: Microsoft-IIS/10.0
WWW-Authenticate: Bearer authorization_uri=https://login.microsoftonline.com/f9daae82-4727-4163-93b9-2214828e3dba, Basic realm="https://tfsprodweu2.app.visualstudio.com/", TFS-Federated
X-TFS-ProcessId: 75e12aa5-f063-480f-9585-cc4044c735c6
Strict-Transport-Security: max-age=31536000; includeSubDomains
ActivityId: 91b05e2e-5142-4345-9785-65d87c166b22
X-TFS-Session: f13f2eea-40f4-42de-a7e2-5875d2a1f39d
X-VSS-E2EID: faa819cb-9214-4623-821f-dd12cd3173e0
X-FRAME-OPTIONS: SAMEORIGIN
X-TFS-FedAuthRealm: https://tfsprodweu2.app.visualstudio.com/
X-TFS-FedAuthIssuer: https://rhjdev01.visualstudio.com/
X-VSS-AuthorizationEndpoint: https://dddev01.vssps.visualstudio.com/
X-VSS-ResourceTenant: f9daae82-4727-4163-93b9-2214828e3dba
X-TFS-SoapException: %3C%3Fxml%20version%3D%221.0%22%20encoding%3D%22utf-8%22%3F%3E%3Csoap%3AEnvelope%20xmlns%3Asoap%3D%22http%3A%2F%2Fwww.w3.org%2F2003%2F05%2Fsoap-envelope%22%3E%3Csoap%3ABody%3E%3Csoap%3AFault%3E%3Csoap%3ACode%3E%3Csoap%3AValue%3Esoap%3AReceiver%3C%2Fsoap%3AValue%3E%3Csoap%3ASubcode%3E%3Csoap%3AValue%3EUnauthorizedRequestException%3C%2Fsoap%3AValue%3E%3C%2Fsoap%3ASubcode%3E%3C%2Fsoap%3ACode%3E%3Csoap%3AReason%3E%3Csoap%3AText%20xml%3Alang%3D%22en%22%3ETF400813%3A%20Resource%20not%20available%20for%20anonymous%20access.%20Client%20authentication%20required.%3C%2Fsoap%3AText%3E%3C%2Fsoap%3AReason%3E%3C%2Fsoap%3AFault%3E%3C%2Fsoap%3ABody%3E%3C%2Fsoap%3AEnvelope%3E

X-TFS-ServiceError: TF400813%3A%20Resource%20not%20available%20for%20anonymous%20access.%20Client%20authentication%20required.

X-VSS-S2STargetService: 00000002-0000-8888-8000-000000000000/visualstudio.com
X-TFS-FedAuthRedirect: https://app.vssps.visualstudio.com/_signin?realm=dddev01.visualstudio.com&reply_to=https%3A%2F%dddev01.visualstudio.com%2F_apis%2Fdistributedtask%2Fpools%2F1%2Fmessages%3FsessionId%3D71e216c8-f6dc-4e56-9648-76d09c99ce6b&redirect=1&context=eyJodCI6MiwiaGlkIjoiYTBhMmQzZDItZTcxNS00YzMyLThmZDctYTVjZjE3MjJkYWU0IiwicXMiOnt9LCJyciI6IiIsInZoIjoiIiwiY3YiOiIiLCJjcyI6IiJ90#ctx=eyJTaWduSW5Db29raWVEb21haW5zIjpbImh0dHBzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbSIsImh0dHBzOi8vbG9naW4ubWljcm9zb2Z0b25saW5lLmNvbSJdfQ2
X-Powered-By: ASP.NET
X-Content-Type-Options: nosniff
[2017-10-11 13:51:12Z INFO RSAEncryptedFileKeyManager] Loading RSA key parameters from file C:\Users\ZAhmadi\Downloads\vsts-agent-win7-x64-2.119.1\.credentials_rsaparams
[2017-10-11 14:05:27Z INFO RSAEncryptedFileKeyManager] Loading RSA key parameters from file C:\Users\ZAhmadi\Downloads\vsts-agent-win7-x64-2.119.1\.credentials_rsaparams
[2017-10-11 14:05:28Z INFO MessageListener] Message '4' received from session '71e216c8-f6dc-4e56-9648-76d09c99ce6b'.
[2017-10-11 14:05:28Z INFO Agent] Refresh message received, kick-off selfupdate background process.

[2017-10-11 14:05:28Z INFO SelfUpdater] Version '2.123.0' of 'agent' package availiable in server.
[2017-10-11 14:05:28Z INFO SelfUpdater] Current running agent version is 2.119.1
[2017-10-11 14:05:28Z INFO SelfUpdater] An update is availiable.
[2017-10-11 14:05:28Z INFO Terminal] WRITE LINE: Agent update in progress, do not shutdown agent.
[2017-10-11 14:05:29Z INFO Terminal] WRITE LINE: Downloading 2.123.0 agent
[2017-10-11 14:05:29Z INFO SelfUpdater] Save latest agent into C:\Users\ZAhmadi\Downloads\vsts-agent-win7-x64-2.119.1\_work\_update\agent.zip.
[2017-10-11 14:07:10Z INFO JobDispatcher] Shutting down JobDispather. Make sure all WorkerDispatcher has finished.

[2017-10-11 14:07:10Z ERR  Terminal] WRITE ERROR: An error occurred: A task was canceled.

[2017-10-11 14:07:10Z ERR  AgentProcess] System.Threading.Tasks.TaskCanceledException: A task was canceled.

   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)


````

**Comment by TingluoHuang** [comment]
@Zaniarr the 401 is by designed, we refresh the oauth token that agent used every hour. 
Can you paste the rest of the log? that seems to be the key part.

**Comment by Zaniarr** [comment]
Logs attached. 
[Agent_20171011-124540-utc.log](https://github.com/Microsoft/vsts-agent/files/1378613/Agent_20171011-124540-utc.log)

**Comment by TingluoHuang** [comment]
@Zaniarr 
base on the stack, looks like it hit http timeout during package download.
do you have internet access to the github agent releases page from the build machine?
the agent try to download the package from `https://github.com/Microsoft/vsts-agent/releases/download/v2.123.0/vsts-agent-win7-x64-2.123.0.zip`
```
[2017-10-11 14:07:10Z ERR  AgentProcess] System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<FinishSendAsync>d__58.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Net.Http.HttpClient.<GetContentAsync>d__33`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Services.Agent.Listener.SelfUpdater.<DownloadLatestAgent>d__10.MoveNext()
```

**Comment by TingluoHuang** [comment]
@Zaniarr anyupdates, also has that agent ever self-updated?

**Comment by Zaniarr** [comment]
Hey, im waiting on the networking team to respond. I've asked them to set up a trace and to investigate why it's timing out. The main guy is on holiday until next week. Will keep you posted

**Comment by TingluoHuang** [comment]
@Zaniarr any updates?

**Comment by Zaniarr** [comment]
Hey, Im still waiting to hear back from my networking team. Im going to go with it's a internal proxy issue for now. You can mark this as resolved. I will open a new issue if/once my networking team come back to me and if i still have the issue. Thanks!!!

**Comment by 199219921992** [comment]
Hello guys, Is the above mentioned issue is resolved? if yes, could you share me the resolution steps ot any workaround for this issue.

**Comment by bryanmacfarlane** [comment]
@199219921992 - in the issue above, an auto update was never invoked and the logged 401 mentioned was a natural part of auth negotiation.  If you have an agent log when this occured, please supply details

**Comment by dogweather** [comment]
A little old, but I'm getting a 401 every time I start up a new runner on ARM / raspberry pi. Although the log file has a 401 auth error, the app outputs _Connected to GitHub ... Listening for Jobs_. However, the runner never shows up in the GitHub web UI. 

It's failing like this both for a repo and my organization. We have 2FA configured.
